xds/internal/xdsclient/transport/transport.go

```diff
@@ -0,0 +1,564 @@
+				resources, url, version, nonce, send = t.processAckRequest(update, stream)
+				if !send {
+					continue
+				}
+				errMsg = update.errMsg
+			}
+			if stream == nil {
+				// There's no stream yet. Skip the request. This request
+				// will be resent to the new streams. If no stream is
+				// created, the watcher will timeout (same as server not
+				// sending response back).
+				continue
+			}
+			if err := t.sendAggregatedDiscoveryServiceRequest(stream, resources, url, version, nonce, errMsg); err != nil {
+				t.logger.Warningf("ADS request for {resources: %q, url: %v, version: %q, nonce: %q} failed: %v", resources, url, version, nonce, err)
+				// Send failed, clear the current stream.
+				stream = nil
+			}
+		}
+	}
+}
+
+// sendExisting sends out xDS requests for existing resources when recovering
+// from a broken stream.
+//
+// We call stream.Send() here with the lock being held. It should be OK to do
+// that here because the stream has just started and Send() usually returns
+// quickly (once it pushes the message onto the transport layer) and is only
+// ever blocked if we don't have enough flow control quota.
+func (t *Transport) sendExisting(stream adsStream) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Reset only the nonces map when the stream restarts.
+	//
+	// xDS spec says the following. See section:
+	// https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version
+	//
+	// Note that the version for a resource type is not a property of an
+	// individual xDS stream but rather a property of the resources themselves. If
+	// the stream becomes broken and the client creates a new stream, the client’s
+	// initial request on the new stream should indicate the most recent version
+	// seen by the client on the previous stream
+	t.nonces = make(map[string]string)
+
+	for url, resources := range t.resources {
+		if err := t.sendAggregatedDiscoveryServiceRequest(stream, mapToSlice(resources), url, t.versions[url], "", ""); err != nil {
+			t.logger.Warningf("ADS request failed: %v", err)
+			return false
+		}
+	}
+
+	return true
+}
+
+// recv receives xDS responses on the provided ADS stream and branches out to
+// message specific handlers.
+func (t *Transport) recv(stream adsStream) bool {
+	msgReceived := false
+	for {
+		resources, url, rVersion, nonce, err := t.recvAggregatedDiscoveryServiceResponse(stream)
+		if err != nil {
+			t.adsStreamErrHandler(err)
+			t.logger.Warningf("ADS stream is closed with error: %v", err)
+			return msgReceived
+		}
+		msgReceived = true
+
+		err = t.updateHandler(ResourceUpdate{
+			Resources: resources,
+			URL:       url,
+			Version:   rVersion,
+		})
+		if e, ok := err.(xdsresource.ErrResourceTypeUnsupported); ok {
+			t.logger.Warningf("%s", e.ErrStr)
+			continue
+		}
+		if err != nil {
+			t.adsRequestCh.Put(&ackRequest{
+				url:     url,
+				version: "",
+				nonce:   nonce,
+				errMsg:  err.Error(),
```

Comment:
Should this field be an `error` instead of a string?

Could we get away with not having this special-cased inside `if err != nil` then (besides the "ack" vs. "nack" logging)?  I'm not sure the significance of not setting `version` here; i.e. is it just not necessary, or must it be unset when nacking, and if the latter can we do that where the request is handled?

Reply:
I switched it to an `error` type. I'm not super sure if this is better though.

As far as getting rid of the `err != nil` is concerned, that is not possible. I added a comment to clarify.

Reply:
> I switched it to an error type. I'm not super sure if this is better though.

My hidden objective was to make the code paths the same so the put/struct creation didn't need to be conditional, but it seems it does need to be.

> As far as getting rid of the err != nil is concerned, that is not possible. I added a comment to clarify.

So...does this mean it was a bug that you weren't setting it before?

Reply:
No, this was being taken care of in `processAckRequest`. But with this change, the latter function is simpler now.

Reply:
I see.  Would it be better to leave that how it was and make this simpler by handling the ack and nack cases together?  Or is it better if the other function is simpler?  (It's probably fine either way.)

Reply:
The reason for having to handle the ack and nack cases separately is more to do with the version number that we have to send and less to do with whether `err != nil`. The latter is used to determine which version number to send. So, I don't think we have completely handle ack and nack cases without some kind of conditional.